feat: Added multi-platform plugin hook support

stevehipwell opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

This PR addresses #7117 to allow plugin installation on Windows in a backwards compatible way. It also adds better OS specific hook support and improved platform command logic to support args containing a space.

I've forked `helm-diff` at https://github.com/stevehipwell/helm-diff to provide an example of how this works,

**Special notes for your reviewer**:

- Closes #7117
- Docs PR https://github.com/helm/helm-www/pull/1627

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

stevehipwell committed:
feat: Added multi-platform plugin hook support

Signed-off-by: Steve Hipwell <steve.hipwell@gmail.com>

stevehipwell commented:
@sabre1041 could you or one of the other maintainers please take a look at this?

sabre1041 commented:
> @sabre1041 could you or one of the other maintainers please take a look at this?

I've started to take a look. Will add some further thoughts in the next day or so

stevehipwell commented:
@sabre1041 have you been able to find any time to look at this? And if so is there any feedback?

sabre1041 commented:
> @sabre1041 have you been able to find any time to look at this? And if so is there any feedback?

getting back up to speed after a work conference. will review this week

sabre1041 reviewed (commented):
Verified in Linux/OSX environments. Will try to find a Windows box to test against this week

stevehipwell commented:
@sabre1041 the error should be consistent, but not necessarily identical, to the existing behaviour on Windows. Basically no currently published Helm plugins can support Windows (unless they have no hooks) so the install script should still fail with the new code.

sabre1041 commented:
> @sabre1041 the error should be consistent, but not necessarily identical, to the existing behaviour on Windows. Basically no currently published Helm plugins can support Windows (unless they have no hooks) so the install script should still fail with the new code.

Contents of PR causes an error while the released code completes successfully. Tested using the [helm diff plugin](https://github.com/databus23/helm-diff)

stevehipwell commented:
@sabre1041 this should now keep the current behaviour on Windows machines which can run `sh` commands.

stevehipwell commented:
@sabre1041 have you had a chance to look at this again since I made the changes to fix the legacy behaviour on a Windows machine with Bash support?

sabre1041 commented:
@stevehipwell shoot, I did review it after you made the prior change and the good news, it works as expected. I'll try to have another maintainer review as well (finding individuals with access to Windows machines could be a challenge)

sabre1041 commented:
@stevehipwell functionality works. However, I am also receiving the same unit test failures that are being reported by the GH run. Are you able to take a look?

stevehipwell commented:
@sabre1041 I've just re-run all of the tests I added or touched and they run fine locally. I've just rebased to see if that was the issue. Could you please re-run the workflow?

sabre1041 commented:
@stevehipwell Still seeing the same issues

stevehipwell commented:
@sabre1041 the issue is due to inconsistent behaviour between the way legacy plugin hooks and legacy plugin commands work. I've updated the code and the tests should pass again now. Could you please re-run them.

stevehipwell commented:
@sabre1041 sorry forgot to push the test changes, could you please re-run again.

stevehipwell commented:
@sabre1041 I've ended up re-writing all of the tests for the plugin prepare logic to be consistent and clear.

The failure on the last version was related to the existing behaviour of taking a mismatching architecture command and using it if there are no full matches. I've removed the test to match the behaviour change I implemented, which doesn't support attempting to use a command for the wrong architecture. I can't see how this is desired behaviour as if the commands are platform specific and the architecture is set this should error. I can see a value to adding support for leaving arch and os empty to allow falling back outside of the legacy `command` value; I haven't implemented this but I could.

sabre1041 reviewed (dismissed):
LGTM

thanks @stevehipwell for working through getting the tests to pass

stevehipwell commented:
@sabre1041 could you point me at instructions for the process to make the corresponding docs update?

sabre1041 commented:
> @sabre1041 could you point me at instructions for the process to make the corresponding docs update?

Docs are found in [this](https://github.com/helm/helm-www).

Doc specific to plugins is [here](https://github.com/helm/helm-www/blob/main/content/en/docs/topics/plugins.md)

stevehipwell commented:
Thanks @sabre1041 I was more referring to the process of updating the docs in sync with code changes?

sabre1041 commented:
> @sabre1041 could you point me at instructions for the process to make the corresponding docs update?

Create a PR against the docs repo and include the reference within this PR

Xitric commented:
Hey @sabre1041, what is the ETA for having this merged? You mentioned looking for another maintainer to review the code.

Our DX team could really benefit from using Helm plugins, but since all of our developers work from Windows platforms, support for hooks on Windows is quite important for us.

stevehipwell commented:
@Xitric I need to add another change to this PR and create a docs PR before it can be re-reviewed. I'm hopeful I can get this done today.

stevehipwell commented:
@sabre1041 this PR is ready for a re-review and I've created the corresponding docs PR (https://github.com/helm/helm-www/pull/1627).

sabre1041 reviewed (commented):
Receiving the following error when running `helm plugin install helm-diff`

```
failed to load plugins: failed to load plugin at "$HELM_PLUGINS/helm-diff/plugin.yaml": error unmarshaling JSON: while decoding JSON: json: unknown field "platformHooks"
```

stevehipwell commented on pkg/plugin/plugin.go:
```diff
@@ -65,23 +66,34 @@ type Metadata struct {
+	// Note that command is not executed in a shell. To do so, we suggest
+	// pointing the command to a shell script.
+	//
+	// DEPRECATED: Use PlatformCommand instead. Remove in Helm 4.
```
This is consistent with other comments.

stevehipwell replied:
@sabre1041 given that you approved the PR can this be resolved?

stevehipwell commented:
The above error looks to be caused by using old code to access my fork of Helm Dif?

sabre1041 commented:
> The above error looks to be caused by using old code to access my fork of Helm Dif?

Indeed I was testing against your fork to verify the configuration. Is it no longer a valid use case?

TBBle commented:
The fork relies on the updated helm from this PR. Unlike the helm-diff fork, I expect most updates will keep the existing `command` stanza for back-compat reasons, and _add_ a `platformCommand` block for updated users, i.e. the hooks block will remain and then a `platfomHooks` block would be added for Windows. Although if existing helm _rejects_ the new `platformHooks` block, then maybe plugin authors can't actually do that?

I also note that in the helm-diff fork, `PlatformCommand` is capitalised, but probably shouldn't be? Same for `PlatformHooks` in the test-case and function comment here. The Go code [correctly specified "platformHooks"](https://github.com/helm/helm/pull/12962/commits/75078fa89d552f8c207c404f2f443febc278bb8f#diff-fc137d62147eb6a9c89c0566479ff219320ca691f2505631ebacbe513bcc8108R119) so I'm mildly surprised this works. YAML is case-sensitive, but maybe the Go yaml parser is not?

Anyway, if possible, I think it might be worth having a test-case for `platformHooks` "no-os" and confirm the expected fallback behaviour to the `hooks` block, even though it's marked to be removed in Helm 4.

stevehipwell commented:
@sabre1041 I created the fork of Helm Diff to only be compatible with the new code, the existing version of Helm Diff should work fine with the new code for currently supported platforms. This was an explicit breaking change for testing purposes, but I wasn't expecting an error about unmarshalling the YAML as the plugin spec has already had `platformCommand` added. I was expecting the new fields to be ignored, is this not the intended behaviour?

@TBBle thanks for spotting the incorrect capitalization, I found one previously and was also confused as to why it worked; these should now be fixed. There is a test in this PR (`TestPrepareCommandsFallback`) to check that the fallback logic works correctly.

TBBle commented:
> I was expecting the new fields to be ignored, is this not the intended behaviour?

Sadly, looks like [we use `UnmarshalStrict`](https://github.com/helm/helm/blob/0b68393d5eba3142063f4cc90298181b7a13379d/pkg/plugin/plugin.go#L231-L233) so that'll fail parsing if it sees unknown fields.

So there's no way to add fields that isn't a back-compat break that I can immediately think of.

stevehipwell commented:
Given that the Helm plugin manifest doesn't follow KRM semantics like the Helm chart manifest and doesn't already support a `helmVersion` field (comparable to the Helm chart `kubernetesVersion` field) I'm don't think there is good solution for this.

Longer term I think the following actions should be considered to augment Helm plugins.

- Switch to `Unmarshal`
- Add an `apiVersion` field for the plugin schema to support breaking changes
- Add a `helmVersion` or `helmMinVersion` field to support minimum version of Helm required to use the plugin

sabre1041 commented:
@stevehipwell I identified the issue from my prior comment. It was due to the plugin hook executing the default `helm` binary on my `PATH` which was a released version and not the version that I was testing as part of the PR, causing the aforementioned error. Once identified and corrected, the subsequent execution ran without issue or warning.

Also agree that we can revisit the structure of plugins in the future.

LGTM

TBBle commented:
> It was due to the plugin hook executing the default `helm` binary on my `PATH` which was a released version and not the version that I was testing as part of the PR, causing the aforementioned error.

Is this something that's worth fixing (elsewhere, not in this PR), perhaps by adding `$(HELM_BINARY)` or similar env-var to `SetupPluginEnv` that plugin script developers can use to get access to the version of Helm used to call their plugin? If supporting `platformHooks` is going to be a back-compat break for the scripts anyway, adding such a feature in the same Helm release would make it immediately usable, and seems like low-hanging fruit.

sabre1041 commented:
> > It was due to the plugin hook executing the default `helm` binary on my `PATH` which was a released version and not the version that I was testing as part of the PR, causing the aforementioned error.
> 
> Is this something that's worth fixing (elsewhere, not in this PR), perhaps by adding `$(HELM_BINARY)` or similar env-var to `SetupPluginEnv` that plugin developers can use to get access to the version of Helm used to call their plugin or its hooks? If supporting `platformHooks` is going to be a back-compat break for the scripts anyway, adding such a feature in the same Helm release would make it immediately usable, and seems like low-hanging fruit.

The issue that I faced was not caused by the core Helm codebase or the contents of the PR, but instead the [hook script](https://github.com/stevehipwell/helm-diff/blob/master/install-binary.sh) that was executed. I, like yourself, worried that there might be adverse effects due to this enhancement. Fortunately, that was not the case.

TBBle commented:
Ah, I see that we have a `HELM_BIN` already, and the issue is that the script [forgot to use it](https://github.com/stevehipwell/helm-diff/blob/master/install-binary.sh#L21C1-L21C98). Oops. That seems trivially fixable, might be worth poking an issue over to https://github.com/databus23/helm-diff

gjenkins8 commented:
Code and spec looks good to me. I did some quick sanity tests, that helm built with changes in this PR can install and run both a new and old plugin:

```
./bin/helm plugin install https://github.com/databus23/helm-diff
./bin/helm diff

# (uninstall)
./bin/helm plugin install https://github.com/stevehipwell/helm-diff
 ./bin/helm diff
```

gjenkins8 reviewed (dismissed):
LGTM but @stevehipwell can we please merge main and fix conflicts (unfortunately)

stevehipwell commented:
@sabre1041 @gjenkins8 I've rebased to `main` and fixed the conflicts.

gjenkins8 commented on pkg/plugin/plugin.go:
```diff
@@ -112,62 +147,101 @@ type Plugin struct {
+// returns the main command and an args array.
 //
 // The result is suitable to pass to exec.Command.
-func (p *Plugin) PrepareCommand(extraArgs []string) (string, []string, error) {
```
I didn't notice this originally, changing the API signature here will limit this change to Helm 4 only (exclude backporting to Helm 3)

gjenkins8 commented on pkg/plugin/plugin.go:
```diff
@@ -112,62 +147,101 @@ type Plugin struct {
+			continue
+		}
+
+		if eq(c.OperatingSystem, runtime.GOOS) || (len(c.OperatingSystem) == 0 && !found) {
```
I think it would be simpler to iterate a third time and do:

```
	for _, c := range cmds {
	 	 if eq(c.OperatingSystem, runtime.GOOS) && c.Architecture == "" {
	 		 return strings.Split(c.Command, " "), c.Args
	 	}
	 }
	 
	for _, c := range cmds {
	 	 if c.OperatingSystem == "" and c.Architecture == "" {
	 		 return strings.Split(c.Command, " "), c.Args
	 	}
	 }
	 
	 return nil, nil
```

rather than overload this loop with the boolean `found`

stevehipwell replied:
I've updated the function for the above comment, but it still uses a single loop. To work correctly with multiple loops we'd need 4 loops (see below) so I think the more complex pattern is valid.

- Exact match
- OS match + empty arch
- Arch match + empty OS
- Empty OS + empty arch

gjenkins8 replied:
👍

I'm not sure if "Arch match + empty OS" makes sense? I don't know of binaries built for a specific arch, that run over all OSs? But support doesn't hurt (aside from the additional, already written, logic)

gjenkins8 replied:
(and I could be wrong)

stevehipwell replied:
I didn't include it in the first place but I added it for completeness. But in this case I'd be happy to lower the complexity and make arch only valid when OS was also set. WDYT?

stevehipwell replied:
I need to make another change so I'll remove the arch only matching logic at the same time.

gjenkins8 commented on pkg/plugin/testdata/plugdir/good/hello/plugin.yaml:
```diff
@@ -3,7 +3,26 @@ version: "0.1.0"
+platformHooks:
+  install:
+    - os: linux
+      arch:
```
```suggestion
```

possible to elide the empty fields? (here and other places)

stevehipwell replied:
Done this one, but I'm not sure of any other instances in the new code.

gjenkins8 reviewed (changes requested):
apologies, I thought I had reviewed this more carefully previously. But a few things jumped out at me when I was re-scanning from the previous changes.

The main items:
- I don't think we should simultaneously support both `Hooks` and `PlatformHooks`, and `Command and `PlatformCommand`. If the user specifies the new field, existance of the old field should be an error.
- I think there are some logic flaws in `PrepareCommands`
- Remove the commented/dead code

stevehipwell commented on pkg/plugin/plugin.go:
```diff
@@ -65,23 +66,34 @@ type Metadata struct {
-	// The following rules will apply to processing commands:
-	// - If platformCommand is present, it will be searched first
+	// The following rules will apply to processing platform commands:
+	// - If PlatformCommand is present, it will be searched first
```
Done

stevehipwell commented on pkg/plugin/plugin.go:
```diff
@@ -65,23 +66,34 @@ type Metadata struct {
 	PlatformCommand []PlatformCommand `json:"platformCommand"`
-	Command         string            `json:"command"`
+
+	// Command is the plugin command, as a single string. The command will be ignored if a valid PlatformCommand is found.
```
I've updated the logic to support this.

gjenkins8 replied:
thanks!

stevehipwell commented on pkg/plugin/plugin.go:
```diff
@@ -90,7 +102,30 @@ type Metadata struct {
+	// pointing the command to a shell script.
+	//
+	// The following rules will apply to processing platform commands:
+	// - If PlatformHooks is present, it will be searched first
```
Done

stevehipwell commented on pkg/plugin/plugin.go:
```diff
@@ -112,62 +147,101 @@ type Plugin struct {
-	var command []string
+// Returns command and args strings based on the following rules in priority order:
+// - From the PlatformCommand where OS and Arch match the current platform
+// - From the PlatformCommand where OS matches the current platform and Arch is empty
```
Done

stevehipwell commented on pkg/plugin/plugin.go:
```diff
@@ -112,62 +147,101 @@ type Plugin struct {
+// Returns command and args strings based on the following rules in priority order:
+// - From the PlatformCommand where OS and Arch match the current platform
+// - From the PlatformCommand where OS matches the current platform and Arch is empty
+// - From the PlatformCommand where OS and Arch are both empty
```
Done

stevehipwell commented on pkg/plugin/plugin.go:
```diff
@@ -112,62 +147,101 @@ type Plugin struct {
+			return strings.Split(c.Command, " "), c.Args
+		}
+
+		if len(c.Architecture) > 0 && !eq(c.Architecture, runtime.GOARCH) {
```
The code would correctly return `linux/arm64` on that platform but I did spot an error where with an empty OS an empty arch could be matched even if there was a more specific arch available depending on order. I've updated this function and tests to fix the issue.

stevehipwell commented:
@gjenkins8 I've addressed your review comments and replied to them. I'll update the parallel docs PR to match the change in spec RE legacy commands.

gjenkins8 reviewed (dismissed):
nice! this is great

stevehipwell commented:
@gjenkins8 @sabre1041 could you please re-review. I've removed the empty OS arch match logic for simplicity.

stevehipwell commented:
@sabre1041 could you please re-review?

sabre1041 commented:
@stevehipwell alright finally getting this in. we have the required approvers. THANK YOU for your patience!

stevehipwell commented:
@sabre1041 @gjenkins8 what's the process for getting the docs PR for this (https://github.com/helm/helm-www/pull/1627) merged?

gjenkins8 commented:
> @sabre1041 @gjenkins8 what's the process for getting the docs PR for this ([helm/helm-www#1627](https://github.com/helm/helm-www/pull/1627)) merged?

We should not forget about the doc PR (thats for creating that). We should aim to merge the docs around release I think.

On that subject, should we backport this to Helm v3 ? (I presume that is wanted).  Currently/now the `main` branch is the dev branch for Helm 4. `dev-v3` needs a backport PR for Helm 3.

stevehipwell commented:
@gjenkins8 the PR was intended for v3, do you need me to do the backport PR or will one of the maintainers do that?

For v4 are there plans to "fix" the plugins? At a minimum by introducing an API version to the plugin schema and supporting marshalling ignoring unknown fields? In addition there doesn't seem to be a consistent plugin lifecycle model, meaning you can install a specific version but after that you're drinking latest from the firehose. And finally, IMHO, the ultimate in plugin architecture would be based on OCI with multi platform images and explicit SemVer versioning; I would image that this could share a large proportion of the OCI chart logic.

stevehipwell commented:
@gjenkins8 @sabre1041 is there an answer to the above question about v3?

sabre1041 commented:
> @gjenkins8 @sabre1041 is there an answer to the above question about v3?

Hey @stevehipwell if you are able to apply the appropriate change to the `dev-v3` branch, it would be appreciated and I'll be happy to review it

yxxhero commented:
@stevehipwell ping.thanks so much.

stevehipwell commented:
@sabre1041 I've been (and still am) mostly AFK. I'll open this PR later on in the week. Do I also need to do the same for the website change?

stevehipwell commented:
I've opened https://github.com/helm/helm/pull/13592 to get these changes into v3. it also looks like the docs were already merged for this change (https://github.com/helm/helm-www/pull/1627).
